Add SGLang Router Support

Bihan opened the pull request:
**Intro**
We want to make it possible to create a gateway which extends the gateway functionality with additional features (all sgl-router features such as cache aware routing, etc) while keeping all the standard gateway features (such as authentication, rate limits).

For the user, using such gateway should be very simple, e.g. setting router to sglang. - for the gateway configurations. The rest for the user should look the same - the same service endpoint, authentication and rate limits working, etc.
While the first change should only bring minimum features - allow to route replicas traffic through the router (dstack’s gateway/ngnix -> sglang-router -> replica workers), in the future this may be extended with router-specific scaling metrics, such as ttft, itl, dissagregated PD. See [TTFT/ITL Based Auto Scaling Issue](https://github.com/dstackai/dstack/issues/3293)).  

As the first experimental version, the most critical is to come up with the minimum changes that are tested thoroughly that would allow embedding the router without breaking any existing functionality.

**Key Changes**
1. Add `src/dstack/_internal/core/models/routers.py`
Define router types and configuration models. The RouterType enum identifies available routers. Each router has its own config model (SGLangRouterConfig, VLLMRouterConfig) with router-specific options. AnyRouterConfig allows to select the correct config class based on the type field.

2. Add `router: AnyRouterConfig`  in `GatewayConfiguration` and  in `GatewayComputeConfiguration`  
Ensure router config flows from user input → server → backend compute layer.

3. Update `gateway/pyproject.toml` to include router packages as optional dependencies

4. Update `get_dstack_gateway_commands()` in `src/dstack/_internal/core/backends/base/compute.py` to accept router config

5. Update `_update_gateway()` in `src/dstack/_internal/server/services/gateways/__init__.py` to extract router_config

6. Add abstract Router base class in `src/dstack/_internal/proxy/gateway/model_routers/base.py`
Handles lifecycle methods of router. 

7. Extend abstract Router base class and implement SGLangRouter in `src/dstack/_internal/proxy/gateway/model_routers/sglang.py`

8. Add router register `src/dstack/_internal/proxy/gateway/model_routers/__init__.py`
Implement the registry pattern (similar to dstack's backend configurators) for auto-discovery and lookup of available routers.

9. Update src/dstack/_internal/proxy/gateway/services/nginx.py

10. Update upstream block of src/dstack/_internal/proxy/gateway/resources/nginx/service.jinja2  to  forward request when router is defined.

11. Add new nginx config to src/dstack/_internal/proxy/gateway/resources/nginx/router_workers.jinja2 to make service replicas's available to TCP port.  Later we could avoid this extra proxying layer by switching from Unix sockets to TCP ports when opening SSH tunnels on the gateway.


**How Router Upgrade Works**

Steps
1. Let’s say Gateway currently has dstack-gateway 0.19.34 and sglang-router 0.2.1 and we are releasing new dstack-gateway 0.19.35 with new SGLang feature in latest SGLang version (sglang-router 0.2.2)
2. We bump up sglang-router version to 0.2.2 in gateway/pyproject.toml
3. dstack server restart -> init_gateways() called -> _update_gateway() -> update.sh executes
4. Flips to inactive venv, let’s say now version = “green”
5. Installs new version in green venv


**Gateway Service Restart and Gateway Instance Reboot**
Router has been tested to successfully reconnect to replicas after both a gateway service restart and a full gateway instance reboot.


**How to test**

Step 1
Apply Below Gateway Config

```
#gateway.dstack.yml
type: gateway
name: bihan-gateway


# Gateways are bound to a specific backend and region
backend: aws
region: eu-west-1

# This domain will be used to access the endpoint
domain: example.com
router:
  type: sglang
  policy: cache_aware
```

Step 2
Update DNS

Step 3
We want to test with multiple services therefore, apply below service configs. 

**Config1**

```
#sglang-service1.yml
type: service
name: sglang-service1

python: 3.12
nvcc: true

env:
  - HF_TOKEN
  - MODEL_ID=meta-llama/Llama-3.2-3B-Instruct
  - HF_HUB_DISABLE_XET=1
  - HF_HUB_ENABLE_HF_TRANSFER=0

commands:
  - pip install --upgrade pip
  - pip install uv
  - uv pip install sglang --prerelease=allow
  - python -m sglang.launch_server --model-path $MODEL_ID --host 0.0.0.0 --port 8000 --enable-metrics

port: 8000
model: sglang-service1.meta-llama/Llama-3.2-3B-Instruct

resources:
  gpu: 24GB

replicas: 0..2
scaling:
   metric: rps
   target: 1
```

**Config2**
```
#sglang-service2.yml
type: service
name: sglang-service2

python: 3.12
nvcc: true

env:
  - HF_TOKEN
  - MODEL_ID=meta-llama/Llama-3.2-3B-Instruct
  - HF_HUB_DISABLE_XET=1
  - HF_HUB_ENABLE_HF_TRANSFER=0

commands:
  - pip install --upgrade pip
  - pip install uv
  - uv pip install sglang --prerelease=allow
  - python -m sglang.launch_server --model-path $MODEL_ID --host 0.0.0.0 --port 8000 --enable-metrics

port: 8000
model: sglang-service2.meta-llama/Llama-3.2-3B-Instruct

resources:
  gpu: 24GB

replicas: 0..2
scaling:
   metric: rps
   target: 1
```

Step 3
To automate request and test autoscaling, you can use below script: `autoscale_test_sglang.py`

```
import asyncio
import aiohttp
import time
import json

# ==== Configuration ====
URL = "https://sglang-service1.example.com/v1/chat/completions" # <-- replace with your endpoint
TOKEN = "esdfds3263-c36d-41db-ba9b-0d31df4efb15e"   # <-- replace with your token
RPS = 2            # requests per second
DURATION = 1800        # duration in seconds
# =======================

HEADERS = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {TOKEN}"
}

PAYLOAD = {
    "model": "sglang-service1.meta-llama/Llama-3.2-3B-Instruct", #<--replace with your model
    "messages": [
        {"role": "system", "content": "You are a helpful assistant."},
        {"role": "user", "content": "What is Deep Learning?"}
    ]
}


async def send_request(session, idx):
    """Send a single request and print full response"""
    try:
        async with session.post(URL, headers=HEADERS, json=PAYLOAD) as resp:
            text = await resp.text()
            print(f"\n[{idx}] Status: {resp.status}")
            print(f"Response:\n{text}\n")
    except Exception as e:
        print(f"[{idx}] Error: {e}")


async def run_load_test():
    total_requests = RPS * DURATION
    interval = 1.0 / RPS

    async with aiohttp.ClientSession() as session:
        start_time = time.perf_counter()
        tasks = []

        for i in range(total_requests):
            tasks.append(asyncio.create_task(send_request(session, i + 1)))
            await asyncio.sleep(interval)

        await asyncio.gather(*tasks)

        elapsed = time.perf_counter() - start_time
        print(f"\n✅ Sent {total_requests} requests in {elapsed:.2f}s "
              f"(~{total_requests/elapsed:.2f} RPS)")


if __name__ == "__main__":
    asyncio.run(run_load_test())
```

Step 6
After updating token and service endpoint, run above script `python autoscale_test_sglang.py ` from your local machine.

Once the automated requests start hitting the service endpoint; dstack submits the job. When the service get's deployed and /health check from sglang-router responds with 200 as shown below, you will start to see response from the model.

As the automated requests continue, first dstack scales up to 2 jobs. If we stop the requests, dstack scales down to 0 jobs.

**Note:** 
1. This PR uses  "https://bihan-test-bucket.s3.eu-west-1.amazonaws.com/dstack_gateway-0.0.1-py3-none-any.whl"  for testing. Later once the PR is ready for merge, I will update it in `src/dstack/_internal/core/backends/base/compute.py`

2. For testing `gateway/pyproject.toml` has my fork as dependency. I will update it once the PR is ready for merge.

Commit:
Add SGLang Router Support

Commit:
Rename router_config to router

Commit:
Rename sglang_workers.jinja2 to router_workers.jinja2

jvstme commented on src/dstack/_internal/server/services/gateways/__init__.py:
```diff
@@ -449,11 +450,24 @@ async def _update_gateway(gateway_compute_model: GatewayComputeModel, build: str
         "rm dstack/_update.sh",
+        # Install gateway package with router extras to the active venv (blue or green)
+        # update.sh writes the active version to dstack/version
+        f"version=$(cat /home/ubuntu/dstack/version) && /home/ubuntu/dstack/$version/bin/pip install --upgrade '{gateway_package}'",
```
Not sure if this works correctly, because the `[sglang]` extra is installed *after* restarting the gateway in `update.sh`, which should mean that the gateway is actually running without the `[sglang]` extra.

You can pass the full package specifier to `update.sh` instead, so that the extra is installed by `update.sh` *before* restarting the gateway.

```shell
$ sh dstack/_update.sh "dstack-gateway[sglang] @ https://bihan-test-bucket.s3.eu-west-1.amazonaws.com/dstack_gateway-0.0.1-py3-none-any.whl"
```

Bihan replied:
Resolved

jvstme commented on src/dstack/_internal/core/models/routers.py:
```diff
@@ -0,0 +1,27 @@
+
+class SGLangRouterConfig(CoreModel):
+    type: Literal["sglang"] = "sglang"
+    policy: str = "cache_aware"
```
TODO: test incorrect values

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/base.py:
```diff
@@ -0,0 +1,147 @@
+            domain: The domain name for this service.
+            num_replicas: The number of replicas to allocate for this domain.
+            model_id: Optional model identifier (e.g., "meta-llama/Meta-Llama-3.1-8B-Instruct").
+                Required only for routers that support IGW (Inference Gateway) mode for multi-model serving.
```
TODO: then why isn't it optional in `Replica`?

Bihan replied:
Resolved. Not required now

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -0,0 +1,366 @@
+                version = "blue"
+
+            # Use Python from the active venv
+            venv_python = DSTACK_DIR_ON_GATEWAY / version / "bin" / "python3"
```
The path to the current interpreter is available in `sys.executable`, no need to reconstruct it from `~/dstack/version`

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -0,0 +1,366 @@
+                capture_output=True,
+                timeout=5,
+            )
+            return result.returncode == 0
```
Calling `curl` looks awkward, and we also can't assume that `curl` is installed on the gateway, since we don't install it explicitly.

Use `httpx` instead.

Same in other `SglangRouter` methods.

Bihan replied:
Resolved

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/base.py:
```diff
@@ -0,0 +1,147 @@
+        self.context = context or RouterContext()
+
+    @abstractmethod
+    def start(self) -> None:
```
`Router.start` and most other `Router` method invocations can take up to several seconds, during which they will block the event loop and prevent the gateway from handling any requests.

To avoid that, you can:
- Make all `Router` methods `async`.
- Revise the `SglangRouter` implementation and replace all blocking calls (`time.sleep`, `subprocess.run`, etc) with non-blocking alternatives (`asyncio.sleep`, `httpx.AsyncClient`, etc).

jvstme commented on src/dstack/_internal/proxy/gateway/services/nginx.py:
```diff
@@ -77,16 +85,47 @@ class Nginx:
                 await run_async(self.run_certbot, conf.domain, acme)
             await run_async(self.write_conf, conf.render(), conf_name)
 
+            if isinstance(conf, ServiceConfig) and conf.router and conf.model_id:
```
> `conf.model_id`

Currently, if you apply a service without `model`, it is provisioned successfully, but doesn't work — the service URL just leads to the router, where this service is not actually registered, so the router always returns 404.

There doesn't seem to be a way for the user to find out that setting `model` is required.

I can suggest add validation for this case on the server (in [`_register_service_in_gateway`](https://github.com/dstackai/dstack/blob/05847cd86e7b66fee2e57a0ec84764abaf2ccf38/src/dstack/_internal/server/services/services/__init__.py#L72)) and reject registering services without `model` on gateways with `router`.

jvstme commented on src/dstack/_internal/proxy/gateway/resources/nginx/service.jinja2:
```diff
@@ -4,9 +4,13 @@ limit_req_zone {{ zone.key }} zone={{ zone.name }}:10m rate={{ zone.rpm }}r/m;
 {% if replicas %}
 upstream {{ domain }}.upstream {
+    {% if router is not none %}
+    server 127.0.0.1:3000;  # SGLang router on the gateway
```
This exposes all SGLang APIs, including `/workers`, to anyone who can request the service.

```shell
> curl https://sglang-service1.my-gateway.example/workers
{"workers":[{"id":"http://127.0.0.1:10006","url":"http://127.0.0.1:10006","model_id":"sglang-service1.meta-llama/Llama-3.2-3B-Instruct","priority":50,"cost":1.0,"worker_type":"regular","is_healthy":true,"load":0,"connection_mode":"Http","metadata":{"model_id":"sglang-service1.meta-llama/Llama-3.2-3B-Instruct"}},{"id":"http://127.0.0.1:10007","url":"http://127.0.0.1:10007","model_id":"sglang-service2.meta-llama/Llama-3.2-3B-Instruct","priority":50,"cost":1.0,"worker_type":"regular","is_healthy":true,"load":0,"connection_mode":"Http","metadata":{"model_id":"sglang-service2.meta-llama/Llama-3.2-3B-Instruct"}}],"total":2,"stats":{"prefill_count":0,"decode_count":0,"regular_count":2}}
```

This might be a security problem for some `dstack` users (especially considering this allows to change workers and not just view them). And I also don't think we want to allow any users to interfere with the worker settings, because they are supposed to be managed by dstack.

I can suggest to forbid requesting `/workers` and other administrative SGLang APIs, if any. Or even introduce a whitelist of allowed routes.

jvstme replied:
Since all service domains just lead to the same router, you can request any service using any domain. For example, in the case from the PR description, you can access the `sglang-service1.meta-llama/Llama-3.2-3B-Instruct` model by requesting `sglang-service2.example.com` (instead of `sglang-service1.example.com`).

This obviously causes various issues:
- It allows unauthorized access to services (if at least one service is `auth: false`, or if we support reusing gateways between projects). 
- It allows to bypass service rate limits.
- RPS autoscaling will be triggered for the service that owns the domain rather than the service that actually handles the load.
- It is confusing.

Not sure how we can address that. If SGLang provides alternatives to routing by model ID (e.g., routing by path or hostname), then we could use them. If not, maybe we could run a separate SGLang router for each service. What do you think?

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/__init__.py:
```diff
@@ -0,0 +1,62 @@
+    _ROUTER_CLASSES.append(SglangRouter)
+    logger.debug("Registered SglangRouter")
+except ImportError as e:
+    logger.warning("SGLang router not available: %s", e)
```
(nit) We can import `SglangRouter` regardless of whether `sglang-router` is installed, so this `except` will never be triggered.

So it seems that this router registration mechanism is redundant (at least in the current version) and can be replaced by a bare `SglangRouter` import.

Bihan replied:
Resolved

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/__init__.py:
```diff
@@ -0,0 +1,62 @@
+        raise ValueError(
+            f"Router type '{router_type.value}' is not available. "
+            f"Available types: {available_types}"
+        )
```
(nit) For consistency with the rest of the `dstack-proxy` codebase, consider using the exceptions from [`proxy.lib.errors`](https://github.com/dstackai/dstack/blob/master/src/dstack/_internal/proxy/lib/errors.py).

Same in a few other `raise` statements in this PR

Bihan replied:
Resolved

Commit:
Resolve SGLang API expose issue

Commit:
Resolve model_id based single router to multi router

Test sglang router per service implementation

Test sglang router per service implementation

Test sglang router per service implementation

Test sglang router per service implementation

Test sglang router per service implementation

Test sglang router per service implementation

Commit:
Resolve gateway installation command

nikita-toffee-ai commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -1021,16 +1022,23 @@ def get_dstack_gateway_wheel(build: str) -> str:
         logger.debug("Found the latest gateway build: %s", build)
-    return f"{base_url}/dstack_gateway-{build}-py3-none-any.whl"
+    # return f"{base_url}/dstack_gateway-{build}-py3-none-any.whl"
+    return "https://bihan-test-bucket.s3.eu-west-1.amazonaws.com/dstack_gateway-0.0.1-py3-none-any.whl"
```
I'm certain it's not supposed to be hard-coded, we need to get the dynamic URL back.

nikita-toffee-ai commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -0,0 +1,269 @@
+                ["lsof", "-ti", f":{self.context.port}"], capture_output=True, timeout=5
+            )
+            if result.returncode == 0:
+                pids = result.stdout.decode().strip().split("\n")
```
Wouldn't a platform-independent `os.linesep` be a better choice here?

Bihan replied:
Here `lsof` always emits "\n" line endings, so splitting on "\n" matches the actual command output.

nikita-toffee-ai commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -0,0 +1,269 @@
+            url = worker.get("url")
+            if url and isinstance(url, str):
+                current_worker_urls.add(url)
+        target_worker_urls = set(replica_urls)
```
We need to make sure there will be no path artifacts in set elements, like trailing slashes, otherwise, the set operations below could yield unexpected results.

nikita-toffee-ai commented on src/dstack/_internal/proxy/gateway/services/nginx.py:
```diff
@@ -164,10 +293,149 @@ def certificate_exists(domain: str) -> bool:
+        )
+
+    def _allocate_worker_ports(self, num_ports: int) -> list[int]:
+        """Allocate worker ports globally in fixed range (10001-11999).
```
I would remove the exact range from the comment, because range ends are vars by nature as we see right down below.

Bihan replied:
Resolved

Bihan commented on src/dstack/_internal/core/backends/base/compute.py:
```diff
@@ -1021,16 +1022,24 @@ def get_dstack_gateway_wheel(build: str) -> str:
         "python3 -m venv /home/ubuntu/dstack/green",
-        f"/home/ubuntu/dstack/blue/bin/pip install {get_dstack_gateway_wheel(build)}",
+        f"/home/ubuntu/dstack/blue/bin/pip install {wheel}",
+        f"/home/ubuntu/dstack/blue/bin/pip install --upgrade '{gateway_package}'",
```
Resolved.

Bihan commented on src/dstack/_internal/core/models/routers.py:
```diff
@@ -0,0 +1,27 @@
+
+class RouterType(str, Enum):
+    SGLANG = "sglang"
+    VLLM = "vllm"
```
Resolved

Commit:
Resolved Minor Review Comments

Commit:
Resolve comments and kubernetes sglang router intregration

Test gateway package update

Test gateway package update

Test gateway package update

Test gateway package update

Test gateway package update

Resolve rate limits and location issue

Resolve rate limits and location issue

Resolve rate limits and location issue

Resolve all major comments

Resolve all major comments

Resolve kubernetes gateway issue with sglang intregration

Bihan commented on src/dstack/_internal/server/services/gateways/__init__.py:
```diff
@@ -454,20 +454,19 @@ async def _update_gateway(gateway_compute_model: GatewayComputeModel, build: str
-        # update.sh writes the active version to dstack/version
-        f"version=$(cat /home/ubuntu/dstack/version) && /home/ubuntu/dstack/$version/bin/pip install --upgrade '{gateway_package}'",
     ]
     stdout = await connection.tunnel.aexec("/bin/sh -c '" + " && ".join(commands) + "'")
```
Resolved

Bihan commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -0,0 +1,269 @@
+
+            venv_python = DSTACK_DIR_ON_GATEWAY / version / "bin" / "python3"
+
+            prometheus_port = self.context.port + 10000
```
Prometheus port is offset by 10000 from router port to keep it in a separate range. Now included in comments

jvstme commented on src/dstack/_internal/proxy/gateway/services/nginx.py:
```diff
@@ -85,46 +85,102 @@ class Nginx:
+                            self._allocated_worker_ports.discard(port)
+
+                    allocated_ports = self._allocate_worker_ports(len(conf.replicas))
+                    self._domain_to_worker_ports[conf.domain] = allocated_ports
```
(nit) This overwrites the old ports in `_domain_to_worker_ports` while they are still in use, which can lead to an inconsistent state if there is an error writing the Nginx config with new ports.

Only update `_domain_to_worker_ports` after writing the Nginx config.

Bihan replied:
Resolved

jvstme commented on src/dstack/_internal/proxy/gateway/services/registry.py:
```diff
@@ -321,13 +329,39 @@ async def get_nginx_service_config(
-                limit_req=LimitReqConfig(zone=zone_name, burst=rate_limit.burst),
+        if is_sglang:
+            for path in sglang_whitelisted_paths:
+                if rate_limit.prefix == path or path.startswith(rate_limit.prefix):
```
(nit) `str_a == str_b` is a special case of `str_b.startswith(str_a)`.


```suggestion
                if path.startswith(rate_limit.prefix):
```

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/__init__.py:
(nit) I'd suggest to move the `model_routers` directory to `proxy/gateway/services/model_routers`.

The subdirectories in `proxy/gateway` are supposed to represent architectural tiers:

- `proxy/gateway/repo` - the data tier
- `proxy/gateway/services` - the logic tier
- `proxy/gateway/routers` and `proxy/gateway/schemas` - the presentation tier

The model routers implementation is part of the logic tier.

Bihan replied:
Resolved

Bihan commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -0,0 +1,255 @@
+                        return True
+                    else:
+                        logger.error("Failed to add worker %s: %s", worker_url, response_data)
+                        return False
```
I've identified an issue affecting SGLang Router versions 0.2.2 and 0.2.3. The worker registration retry mechanism stops at 10 attempts regardless of the --worker-startup-timeout-secs setting. Version 0.2.2 was working until recently. I'm investigating the root cause and will provide an update once I have more details.

Bihan commented on src/dstack/_internal/proxy/gateway/model_routers/base.py:
```diff
@@ -0,0 +1,91 @@
+
+    host: str = "127.0.0.1"
+    port: int = 3000
+    log_dir: Path = Path("./router_logs")
```
Yes router_port is only set when a router exists and only sglang router exists currently.

Bihan replied:
Resolved

jvstme commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -0,0 +1,255 @@
+                        return False
+                else:
+                    logger.error(
+                        "Failed to remove worker %s: status %d, %s",
```
(nit) I'm seeing this error in `systemctl -u dstack.gateway`:

```
dstack._internal.proxy.gateway.model_routers.sglang - ERROR - Failed to remove worker http://127.0.0.1:10003: status 202, {"status":"accepted","worker_id":"http://127.0.0.1:10003","message":"Worker removal queued for background processing"}
```

To reproduce:
- Start a service with one replica.
- Scale it to two replicas.

Bihan replied:
Resolved. This was happening with all Add/Remove call. This happened because I was expecting status 200, while Add/Remove worker apis responded with 202.

Commit:
Resolve additional comments

Bihan commented on src/dstack/_internal/proxy/gateway/resources/nginx/service.jinja2:
```diff
@@ -4,9 +4,15 @@ limit_req_zone {{ zone.key }} zone={{ zone.name }}:10m rate={{ zone.rpm }}r/m;
 {% if replicas %}
 upstream {{ domain }}.upstream {
+    {% if router is not none %}
+    {% if router.type == "sglang" and router_port is not none %}
```
Resolved

Bihan commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -1,366 +1,255 @@
             else:
-                logger.debug("No sglang-router process found to stop")
+                result = subprocess.run(
+                    ["pgrep", "-f", f"sglang.*--port.*{self.context.port}"],
```
Resolved

Bihan commented on src/dstack/_internal/proxy/gateway/services/nginx.py:
```diff
@@ -213,6 +290,125 @@ def certificate_exists(domain: str) -> bool:
+                    # If bind fails (e.g., Address already in use), port is not available
+                    return False
+        except Exception:
+            logger.debug("Error checking port %s availability", port)
```
Resolved

Bihan commented on src/dstack/_internal/proxy/gateway/services/nginx.py:
```diff
@@ -85,46 +85,102 @@ class Nginx:
+                        for port in allocated_ports:
+                            self._allocated_worker_ports.discard(port)
+                        if conf.domain in self._domain_to_worker_ports:
+                            del self._domain_to_worker_ports[conf.domain]
```
Resolved

Commit:
Pinned sglang-router to 0.2.1

Bihan committed:
Merge branch 'master' into add_sglang_router_support

Commit:
Fix linting error

Bihan commented on src/dstack/_internal/proxy/gateway/model_routers/sglang.py:
```diff
@@ -1,366 +1,255 @@
+                            pid,
+                            self.context.port,
+                        )
+                        subprocess.run(["kill", pid], timeout=5)
```
Resolved
